Use webpacker

himdel opened the pull request:
~~Depends on https://github.com/ManageIQ/manageiq/pull/15054~~
~~Depends on https://github.com/ManageIQ/manageiq/pull/15578~~

This adds [webpacker](https://github.com/rails/webpacker) support to ui-classic. Webpacker reads `app/javascript` from all rails engines, and outputs bundled assets to `manageiq/public/packs/`.


##### This adds the following rake tasks (runnable from `manageiq/`):

   * `webpack:compile` - runs `webpacker:compile`, which runs webpack once, producing assets in `manageiq/public/packs/` (called on `assets:precompile`, and `bin/update`)
      * `webpack:clobber` - cleans the generates packs

   * ~~`webpack:server` - runs the webserver - presumably mostly for UI developers, use instead of `bin/webpack-dev-server` please~~ (webpack server is problematic with CORS, use at your own risk)


##### Expected workflow changes:

   * non-UI devs: do not care, running `bin/update` runs webpack for you, so unless you're doing UI changes, or not running `bin/update` after changing branches, you're fine. A rebuilt can be still triggered manually via `rake webpack:compile` (or `rake update:ui` to also get dependencies)

   * UI devs: if you're doing any development in `app/javascript/` you don't want to have to run `rake webpack:compile` all the time, instead, you can run `bin/webpack --watch` (in the ui-classic root, once you have run the rake task at least once) - this will watch for any changes and automatically recompile.


##### Internally (runnable only from `manageiq-ui-classic/`), we have a few more tasks:

   * `webpack:paths` - outputs `Rails.root`, and a JSON hash of `engine class => absolute path to that engine's root` to `config/webpack/paths.json` - needed by the JS bits to locate all the engines and to know where to find that `public/` to output to

   * `yarn:install` - this is a Rails 5.1 task not available in Rails 5.0. webpacker tries to do too much when that task does not exist, ~~so we're redifining it to just call the ol' `update:yarn`~~ *and* it still tries to run this 6 times during `webpack:compile`, so we're defining it to do nothing, and calling `update:yarn` is handled by `bin/update`

   * we also have all the default webpacker tasks we're wrapping, including `webpacker:compile`, `webpacker:clobber`, and also `webpacker:verify_install`, `webpacker:check_node`, `webpacker:check_yarn`, needed by the other tasks


##### Other magics:

   * our `config/initializer/webpack.rb` needs to override all the `*path` methods defined in the gem's `./lib/webpacker/configuration.rb`, except those dealing with *output*
   * our `config/webpack/configuration.js` reads the `webpack:paths` output to determine `Rails.root`, so that the output goes to manageiq, and to see all the other engines' roots
   * out rake tasks def requires the webpacker initializer - becasue we need the paths overriden even in rake tasks (well, especially in those)
   * the choice to *wrap* the webpacker tasks instead of exposing them directly was done because I couldn't get it to work quite right from `manageiq/` without hackery like running `Webpacker.bootstrap`, but only when in that directory (not sure if there's a better way, and really only assuming the problem has to do with engine initialization ordering)


##### TODO:

   * [x] upgrade to webpacker 2.0
   * [x] rake hooks
   * [x] read from all the rails engines' roots
   * [x] don't webpack bower_components, etc (fixed by 2.0, and manually skipping `spec/manageiq` (because of SUI bits in `public/ui`))
   * [x] bin/update changes - run `update:yarn`, and `webpack:compile`
   * [ ] appliance build changes (separate PR)
   * [x] ~~switching between dev-server and compile?~~ - problematic now, use `webpack --watch`
   * [x] go through the other TODOs :metal:-l
   * [x] npm from all the plugins - but install to plugin dir, and when webpacking, read from there (and maybe also ui-classic?)
   * [ ] (later - config changes) https://github.com/theforeman/foreman/blob/develop/config/webpack.config.js#L124

##### Notes:

When Rails 5.1 comes out, we'll probably need to make `update:yarn` to call that, and not redefine it to go the other way.

When we want to add react support, `webpacker-2.0/lib/install/react.rb` should help quite a bit :).

JS living under `app/assets/javascripts/` is **not** affected in any way. JS living under `app/javascript` can be made into packs, which can be included from hamls via `javascript_pack_tag`.

abonas commented:
cc @mtho11 @karelhala

karelhala commented:
@himdel I love how you split webpack into multiple files (`sass.js`, `coffee.js`) etc.

I know that it's still in WIP phase, but consider moving couple of `dependencies` into `devDependencies`? for instance most of the loaders and webpack plugins are not quite dependencies, but development dependencies, doesn't it?

himdel commented:
@karelhala just a note, so far, I've done nothing, this is just the default webpacker config.. :)

But yeah, we definitely want to separate the build deps from the actual ui deps, will do :)

guilleiguaran commented:
Hi there, I'm happy to review this and check what we can do in upstream webpacker to get it working correctly with engines 🙌🏻

himdel commented:
@guilleiguaran Thanks, I've updated the description to match the current state more closely.. and about to update the PR to 2.0, hopefully :)

ohadlevy commented:
@guilleiguaran thats awesome, we would love to have that support on the https://github.com/theforeman/foreman - we use engines a lot and also use webpack, where is the right way to discuss this further? :)

himdel commented:
So, this is almost ready, `be rake webpack:compile` works from both repos, and produces compiled packs without errors.

The typescript config is still a bit wrong, looks like it tries to read any `*.ts` files anywhere, not only from `app/javascript`.

`be rake webpack:server` works from both repos too, but still tries to run yarn when run from manageiq, for some reason.

So except for appliance build and `bin/update` integrations, the only thing that remains to be resolved here is serving the packs to UI devels without requiring non-UI devels to run `webpack:server`.

himdel commented:
Ad CSP - in `manageiq/`

```diff
diff --git a/config/initializers/secure_headers.rb b/config/initializers/secure_headers.rb
index fbbab5da1b..7e89e7b4f3 100644
--- a/config/initializers/secure_headers.rb
+++ b/config/initializers/secure_headers.rb
@@ -1,4 +1,6 @@
 SecureHeaders::Configuration.default do |config|
+  allow_webpacker = Rails.env.development? ? ['0.0.0.0:8080', 'localhost:8080', '127.0.0.1:8080'] : []
+
   config.hsts = "max-age=#{20.years.to_i}"
   config.x_frame_options = 'SAMEORIGIN'
   config.x_content_type_options = "nosniff"
@@ -7,9 +9,9 @@ SecureHeaders::Configuration.default do |config|
     :report_only => false,
     :default_src => ["'self'"],
     :frame_src   => ["'self'"],
-    :connect_src => ["'self'"],
+    :connect_src => ["'self'", *allow_webpacker],
     :style_src   => ["'unsafe-inline'", "'self'"],
-    :script_src  => ["'unsafe-eval'", "'unsafe-inline'", "'self'"],
+    :script_src  => ["'unsafe-eval'", "'unsafe-inline'", "'self'", *allow_webpacker],
     :report_uri  => ["/dashboard/csp_report"]
   }
 end
```

almost works, except the connect_src value gets rewritten by `ApplicationController#allow_websocket` - need to make it aware of webpacker as well, or figure out how to add stuff instead of replacing.

himdel commented on config/webpack/loaders/erb.js:
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  test: /\.erb$/,
```
Webpacker default :).

Realistically, I think we have zero use for `erb` here, but we probably will need to do a `haml` pack - right now, every angular directive/component template (anything referenced by `templateUrl`) lives in `app/views/static/` and we have a special controller that just serves those, passing the template through gettext and haml->html conversions ... but in the future, we should probably compile those on the server and do per-language template bundles..

himdel replied:
I didn't really want to change the default config in the same PR, since there are already some changes to those to just make it work.

But we will need to clean those defaults a bit, agreed there.. :)

ohadlevy commented on package.json:
```diff
@@ -16,17 +16,57 @@
+    "style-loader": "~0.16.1",
+    "ts-loader": "~2.0.3",
+    "typescript": "~2.2.2",
+    "webpack": "~2.4.1",
```
Would you consider using 2.6.2 ? (as its seems the latest release of the 2.x branch)?

himdel replied:
Definitely, assuming there's no compatibility issues, will do :)

mtho11 replied:
Webpack 3.0 is supposed to be backward compatible. Might be worth a test.

mtho11 replied:
Let's get this ready for everyone. Invite whoever to download and try. I know I had many little issues that cost me time.

mtho11 replied:
Perhaps an 'angular2' branch or whatever that various parties can try out?

mtho11 replied:
We need to get this to mainstream now.

mtho11 replied:
Perhaps an `angular.next` branch is preferable?

mtho11 replied:
How does an individual know everything is setup properly? 
I did not see any instructions on verifying installation?

himdel replied:
Right now, verifying instalation means you run `rake webpack:compile`, and open the web browser console in the UI, and you should see a `Hello World from Webpacker` message.

I see no reason to create a separate branch for testing this, the branch already exists. And once this is in, we can use the master branch for angular 2. So.. what exactly was the idea behing that branch?

himdel commented:
@mtho11 I've updated the PR to fix that bundler error you were seeing.

Before, we were running `webpack:compile`, which ran `webpacker:compile` which ran webpack, which ran `webpack:output` and `webpack:engine`.

Now, running `webpack:compile` first runs `webpack:paths` which generates `config/webpack/paths.json`, and the runs `webpack:compile` and webpack only reads that config file.

The rest of the pending changes is coming shortly, but wanted to push this so you can test the problem is gone, thanks :).

himdel commented:
So this is pretty much ready. 

Still needs documentation, an appliance build and pods change, but.. it works in the sense that running `rake webpack:compile` compiles `app/javascripts` without failing :).

To test.. pull this PR, run `bin/update`, and run manageiq.. In the browser console you should see a `'Hello World from Webpacker'` message.

mtho11 commented:
@himdel my old version works (1 week+) the new version doesn't. Here are the cmd outputs:

[Old version](https://paste.fedoraproject.org/paste/kSsuas3ukTusqUAMqgbouA)

[Latest version](https://paste.fedoraproject.org/paste/afOacrNsmXbDIydw1PVNpw)

mtho11 commented:
@karelhala can you try out as well?

himdel commented:
Sounds like webpacker itself works, that's good news, but the typescript error is .. weird. 

The good news is that we can probably ignore those, but.. yeah, I'd like to get rid of them.

Still can't reproduce though, I assume you ran `yarn` before? Or try `yarn upgrade`...

(Also, does this depend on where you run the rake task from? It should not, but just to be sure...)
(Aaand, please check that your `config/webpack/paths.json` only has ui-classic in `engines`, to make sure no other plugin's dependencies are involved.)

mtho11 commented:
@himdel I did run `yarn` and tried again with `yarn upgrade` same results.
Good to know about the `engines` plugins piece though.

mtho11 commented:
I think the typescript compiler errors are fine but the `webpacker:compile` task should not fail:

`[Webpacker] Compilation Failed`
`== webpacker:compile failed ==`

Perhaps, I'm the only one seeing this. Hopefully, we can get some other testers to share their validations here for a sanity check? (I'm already certifiably insane, but that is beside the point).

himdel commented:
@mtho11 What's funny is that running `be rake webpack:compile`, I still get no errors.. 

But running `bin/webpack` does give me..

```
WARNING in ./node_modules/@angular/core/@angular/core.es5.js
5870:15-36 Critical dependency: the request of a dependency is an expression
    at ImportLazyContextDependency.getWarnings (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/dependencies/ImportContextDependency.js:28:4)
    at Compilation.reportDependencyErrorsAndWarnings (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:676:24)
    at Compilation.finish (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:534:9)
    at applyPluginsParallel.err (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compiler.js:506:17)
    at /home/himdel/manageiq-ui-classic/node_modules/tapable/lib/Tapable.js:271:11
    at _addModuleChain (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:480:11)
    at processModuleDependencies.err (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:451:13)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)

WARNING in ./node_modules/@angular/core/@angular/core.es5.js
5886:15-102 Critical dependency: the request of a dependency is an expression
    at ImportLazyContextDependency.getWarnings (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/dependencies/ImportContextDependency.js:28:4)
    at Compilation.reportDependencyErrorsAndWarnings (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:676:24)
    at Compilation.finish (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:534:9)
    at applyPluginsParallel.err (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compiler.js:506:17)
    at /home/himdel/manageiq-ui-classic/node_modules/tapable/lib/Tapable.js:271:11
    at _addModuleChain (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:480:11)
    at processModuleDependencies.err (/home/himdel/manageiq-ui-classic/node_modules/webpack/lib/Compilation.js:451:13)
    at _combinedTickCallback (internal/process/next_tick.js:73:7)
    at process._tickCallback (internal/process/next_tick.js:104:9)
```

Not the errors you were seeing, but yes the warnings.

himdel commented:
.. and the warnings should be fixed.

The issue is https://github.com/angular/angular/issues/11580, the fix comes from  https://angular.io/guide/webpack#common-configuration .

himdel commented:
TODO - travis - don't forget to run `webpack:compile` :)

martinpovolny commented:
@himdel : please, replace

mtho11 commented:
Can we also add an example Haml page exercising ng4 output(hello-world) in some sample screen? I'm not able to get this to work in current iteration. Although I do see the console.log message from application.js no output to a screen is possible (for me).

himdel commented:
@mtho11 I'm not sure for what kind of example you're asking...

If I do

```diff
--- a/app/views/dashboard/show.html.haml
+++ b/app/views/dashboard/show.html.haml
@@ -19,3 +19,7 @@
 - if WidgetPresenter.chart_data.present?
   :javascript
     ManageIQ.charts.chartData = #{{"widgetchart" => WidgetPresenter.chart_data}.to_json.html_safe};
+
+%hello-angular
+  Loading...
+= javascript_pack_tag 'hello_angular'
```

I can see the angular2 component running below the dashboard. Anything else is way out of scope of this PR..

himdel commented:
Rebased, fixed travis, and updated the description to reflect the current state, and to mention `bin/webpack --watch`.

Should be ready :) (famous last words..)

himdel commented:
Note.. @mtho11 noticed `javascript_pack_tag` can't be used from any partial which gets ajax-loaded. That's because it just creates a `<script>` tag and when such a script tag is added via jquery, it triggers a load which tries to do the same as every other jquery load .. tries to parse the response as a JSON, and then tries to strip the `throw "error"` thing and eval it.. neither of which works.

But the idea is to include the packs from `application.html.haml` anyway, where there's no problem.

martinpovolny commented:
@h-kataria : can you, please, test this one?

h-kataria commented:
@martinpovolny @dclarizio Tried it in UI, good to be merged

ohadlevy commented on package.json:
```diff
@@ -16,20 +16,58 @@
+    "webpack-merge": "~4.1.0"
   },
   "engines": {
     "node": ">= 6.9.1",
```
maybe not relevant, but I'm getting this when running webpack:

```
warning package.json: License should be a valid SPDX license expression
```

himdel replied:
It's probably complaining about a dependency?

We have `"license": "Apache-2.0",`, which would seem to match perfectly with https://spdx.org/licenses/Apache-2.0

himdel commented:
@ohadlevy you shouldn't need to run the webpacker installer locally, all the files are here, so just `yarn`.

After that, `rake webpack:compile` works, and after that, `bin/webpack --watch` does too.
